### PR TITLE
Replace list with generator when iterating headers

### PR DESCRIPTION
**Pull request by mvdbeek**
This should be more memory-friendly. @blankenberg suggested doing this
in #4319.

### CODE REVIEW

File: lib/galaxy/datatypes/interval.py [hunk 1]
```diff
@@ -313,7 +316,7 @@ def sniff( self, filename ):
         >>> Interval().sniff( fname )
         True
         """
-        headers = get_headers( filename, '\t', comment_designator='#' )
+        headers = iter_headers( filename, '\t', comment_designator='#' )
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/interval.py]: I'd move the `iter_headers()` call inside the `try` (i.e. `for hdr in iter_headers( filename, '\t', comment_designator='#' )` ) and remove this line. Similarly below.

File: lib/galaxy/datatypes/mothur.py [hunk 2]
```diff
@@ -32,10 +35,11 @@ def set_meta(self, dataset, overwrite=True, **kwd):
...
-            headers = get_headers(dataset.file_name, sep='\t', count=-1)
+            headers = iter_headers(dataset.file_name, sep='\t', count=-1)
+            first_line = next(headers)
             # set otulabels
-            if len(headers[0]) > 2:
-                otulabel_names = headers[0][2:]
+            if len(first_line) > 2:
+                otulabel_names = first_line[2:]
             # set label names and number of lines
             for line in headers:
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/mothur.py:44]: Here you are not processing the first line any more because of the `next()` call.

Comment (mvdbeek) [reply lib/galaxy/datatypes/mothur.py:44]: I switched to fetching the first line separately in 43702ccacdbfe925d921dac413c02bb7f65b9c4f

**Review by nsoranzo — APPROVED** [review]

**Comment by jmchilton** [comment]
Beautiful - thanks!

For what it is worth, I believe the amount of memory can still grow unbounded because of this line:

```
for line in in_file:
```

It should be replaced by something that reads in lines not exceeding 10 Kb or something like that.

In fact - I'd take caching that first 10 Kb and not using the iterator over re-reading from the disk over and over. Caching the headers and using an iterator though is the best combination I suspect so this is still a very solid step in the right direction.

**Comment by blankenberg** [comment]
+1 Thanks @mvdbeek 

As @jmchilton mentioned, very long lines can still consume unbounded memory, but that is a more difficult issue to deal with in a backwards compatible manner, as it could impact dataset metadata for certain datatypes and end up changing tool behavior.

**Comment by jmchilton** [comment]
> but that is a more difficult issue to deal with in a backwards compatible manner, as it could impact dataset metadata for certain datatypes and end up changing tool behavior

Isn't that a separate problem from constraining sniffing's memory usage?

**Comment by blankenberg** [comment]
> Isn't that a separate problem from constraining sniffing's memory usage?

In general, yes. But I was referring to any nonparameterized changes to the x_headers() methods.

**Comment by jmchilton** [comment]
> In general, yes. But I was referring to any nonparameterized changes to the x_headers() methods.

Gotcha - thanks!

**Pull request merged**